torch/fx/_proxy_tensor.py

```diff
@@ -0,0 +1,201 @@
+        if func.__name__[-1] == "_" and func.__name__[0] != "_":
+            args[0].proxy = proxy_out
+            proxy_out.node.meta['tensor_meta'] = _extract_tensor_metadata(args[0])
+
+        with no_dispatch():
+            real_out = func_overload(*args, **kwargs)
+
+        def wrap_with_proxy(e, proxy):
+            # Some ops (like native_batch_norm_backward) return undefined tensors that get
+            # converted into None in python.
+            # As the function signature expects tensors, if we directly return these None
+            # tensors back to C++, we'll error.
+            if e is None:
+                e = torch.empty(())
+            if type(e) == torch.Tensor:
+                return ProxyTensor(e, proxy)
+            else:
+                return e
+        if isinstance(real_out, tuple):
+            return tuple([wrap_with_proxy(e, proxy_out[idx]) for idx, e in enumerate(real_out)])
+        elif isinstance(real_out, list):
+            return list([wrap_with_proxy(e, proxy_out[idx]) for idx, e in enumerate(real_out)])
+        elif isinstance(real_out, torch.Tensor):
+            return wrap_with_proxy(real_out, proxy_out)
+        else:
+            return real_out
+
+
+class PythonKeyTracer(Tracer):
+    def __init__(self):
+        super().__init__()
+
+    def call_module(
+        self, m: torch.nn.Module, forward: Callable[..., Any], args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Any:
+        return forward(*args, **kwargs)
+
+    def _module_getattr(self, attr, attr_val, parameter_proxy_cache):
+        if isinstance(attr_val, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if attr_val is p:
+                    if n not in parameter_proxy_cache:
+                        proxy = self.create_proxy('get_attr', n, (), {})
+                        parameter_proxy_cache[n] = ProxyTensor(attr_val, proxy)
+                    return parameter_proxy_cache[n]
+            return attr_val
+        return attr_val
+
+    # We need to do this so that parameters entering the `make_fx` context have
+    # a reference to them (and also have requires_grad set on them correctly
+    # I'm not actually sure if this is the right thing to do ...
+    def create_arg(self, a: Any):
+        if isinstance(a, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if a is p:
+                    return self.create_node('get_attr', n, (), {})
+            qualname: Optional[str] = None
+
+            if not qualname:
+                i = 0
+                while True:
+                    qualname = f'_param_constant{i}'
+                    if not hasattr(self.root, qualname):
+                        break
+                    i += 1
+                setattr(self.root, qualname, a)
+
+            return self.create_node('get_attr', qualname, (), {})
+        return super().create_arg(a)
```

Comment:
This is the part of Horace's implementation that I'd most like to get eyeballs from FX folks on, idk maybe @jamesr66a @zdevito ?

Reply:
So is what's going on here that in a ProxyTensor context, you have `nn.Parameter` instances that are not installed as members of the module hierarchy and you'd like to intern those with a generated qualified name? Do you have an example of where this could happen?

Reply:
yes.

So, if you're tracing something like

```
model = ...
def f(x):
    return model(x)
```
So, I think it's natural that in this case, the model parameters should just be constants from the perspective of the tracer. But since FX overrides parameters to look them up on the module hierarchy, this behavior needs to be modified.

Reply:
I talked with @jamesr66a and we think some sort of approach along these lines is reasonable.

James has a preference for preserving as much of FX's current hooks as possible.